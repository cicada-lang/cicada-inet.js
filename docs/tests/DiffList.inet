deftype Trivial 0 end

defnode sole
  ------
  value!: Trivial
end

deftype List 1 end

defnode null
  ------
  value!: List('a)
end

defnode cons
  head: 'a
  tail: List('a)
  -------
  value!: List('a)
end

deftype DiffList 1 end

defnode diff
  front: List('a)
  ------
  back: List('a)
  value!: DiffList('a)
end

// Concatenation of lists is performed in linear time
// with respect to its first argument.
// Constant time concatenation is possible
// with difference-lists: the idea consists in
// plugging the front of the second argument
// at the back of the first one.

defnode diff_append
  target!: DiffList('a)
  rest: DiffList('a)
  ------
  return: DiffList('a)
end

defnode diff_open
  target!: DiffList('a)
  list: List('a)
  ------
  return: List('a)
end

defrule diff diff_append
  (diff)-front diff return-(diff_append)
  (diff_append)-rest diff_open back-(diff)
end

defrule diff diff_open
  (diff)-back return-(diff_open)
  (diff)-front list-(diff_open)
end

defn one_two_soles
  wire sole cons diff rot rot connect
  wire sole cons sole cons diff rot rot connect
  diff_append
end

show one_two_soles end
run one_two_soles end

defn two_two_soles
  wire sole cons sole cons diff rot rot connect
  wire sole cons sole cons diff rot rot connect
  diff_append
end

show two_two_soles end
run two_two_soles end

defn two_two_soles_with_local_variable
  wire sole cons sole cons diff $value connect value
  wire sole cons sole cons diff $value connect value
  diff_append
end

show two_two_soles_with_local_variable end
run two_two_soles_with_local_variable end
