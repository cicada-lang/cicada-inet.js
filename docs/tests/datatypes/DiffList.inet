require "List.inet"

type DiffList Type -- Type end

node diff
  'A List :front
  -------
  'A List :back
  'A DiffList :value!
end

// Concatenation of lists is performed in linear time
// with respect to its first argument.
// Constant time concatenation is possible
// with difference-lists: the idea consists in
// plugging the front of the second argument
// at the back of the first one.

node diff_append
  'A DiffList :target!
  'A DiffList :rest
  --------
  'A DiffList :return
end

node diff_open
  'A DiffList :target!
  'A List :list
  ----------
  'A List :return
end

rule diff diff_append
  (diff)-front diff return-(diff_append)
  (diff_append)-rest diff_open back-(diff)
end

rule diff diff_open
  (diff)-back list-(diff_open)
  (diff)-front return-(diff_open)
end
