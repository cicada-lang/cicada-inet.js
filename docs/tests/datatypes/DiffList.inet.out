(diff₁)-value!target-(diff_append₀)
(diff₁)-back tail-(cons₁)
(diff₀)-back tail-(cons₀)
(diff₀)-value rest-(diff_append₀)
(cons₂)-value front-(diff₁)
(cons₁)-value tail-(cons₂)
(sole₂)-value head-(cons₂)
(sole₁)-value head-(cons₁)
(cons₀)-value front-(diff₀)
(sole₀)-value head-(cons₀)
(diff_append₀)-return

(cons₀)-value tail-(cons₁)
(cons₀)-tail back-(diff₂)
(cons₂)-value front-(diff₂)
(cons₁)-value tail-(cons₂)
(sole₂)-value head-(cons₂)
(sole₁)-value head-(cons₁)
(sole₀)-value head-(cons₀)
(diff₂)-value

(diff₄)-value!target-(diff_append₁)
(diff₄)-back tail-(cons₅)
(diff₃)-back tail-(cons₃)
(diff₃)-value rest-(diff_append₁)
(cons₆)-value front-(diff₄)
(cons₅)-value tail-(cons₆)
(sole₆)-value head-(cons₆)
(sole₅)-value head-(cons₅)
(cons₄)-value front-(diff₃)
(cons₃)-value tail-(cons₄)
(sole₄)-value head-(cons₄)
(sole₃)-value head-(cons₃)
(diff_append₁)-return

(cons₄)-value tail-(cons₅)
(cons₃)-tail back-(diff₅)
(cons₆)-value front-(diff₅)
(cons₅)-value tail-(cons₆)
(sole₆)-value head-(cons₆)
(sole₅)-value head-(cons₅)
(cons₃)-value tail-(cons₄)
(sole₄)-value head-(cons₄)
(sole₃)-value head-(cons₃)
(diff₅)-value

(diff₇)-value!target-(diff_append₂)
(diff₇)-back tail-(cons₉)
(diff₆)-back tail-(cons₇)
(diff₆)-value rest-(diff_append₂)
(cons₁₀)-value front-(diff₇)
(cons₉)-value tail-(cons₁₀)
(sole₁₀)-value head-(cons₁₀)
(sole₉)-value head-(cons₉)
(cons₈)-value front-(diff₆)
(cons₇)-value tail-(cons₈)
(sole₈)-value head-(cons₈)
(sole₇)-value head-(cons₇)
(diff_append₂)-return

(cons₈)-value tail-(cons₉)
(cons₇)-tail back-(diff₈)
(cons₁₀)-value front-(diff₈)
(cons₉)-value tail-(cons₁₀)
(sole₁₀)-value head-(cons₁₀)
(sole₉)-value head-(cons₉)
(cons₇)-value tail-(cons₈)
(sole₈)-value head-(cons₈)
(sole₇)-value head-(cons₇)
(diff₈)-value

